modules/table/src/main/java/org/apache/ignite/internal/table/distributed/disaster/GroupUpdateRequestHandler.java

```diff
@@ -0,0 +1,676 @@
+     * Adds more nodes into {@code partAssignments} until it matches the number of replicas or we run out of nodes.
+     */
+    private static void enrichAssignments(
+            PartitionGroupId partId,
+            Collection<String> aliveDataNodes,
+            int partitions,
+            int replicas,
+            Set<Assignment> partAssignments
+    ) {
+        Set<Assignment> calcAssignments = calculateAssignmentForPartition(aliveDataNodes, partId.partitionId(), partitions, replicas);
+
+        for (Assignment calcAssignment : calcAssignments) {
+            if (partAssignments.size() == replicas) {
+                break;
+            }
+
+            partAssignments.add(calcAssignment);
+        }
+    }
+
+    abstract CompletableFuture<Map<T, LocalPartitionStateMessageByNode>> localStatesFuture(
+            DisasterRecoveryManager disasterRecoveryManager,
+            Set<String> zoneNames,
+            Set<Integer> partitionIds,
+            Catalog catalog
+    );
+
+    abstract CompletableFuture<Map<Integer, Assignments>> stableAssignments(
+            MetaStorageManager metaStorageManager,
+            int zoneId,
+            int[] partitionIds
+    );
+
+    abstract T replicationGroupId(int id, int partitionId);
+
+    abstract CompletableFuture<Integer> invoke(
+            T partId,
+            long revision,
+            HybridTimestamp timestamp,
+            MetaStorageManager metaStorageMgr,
+            long assignmentsTimestamp,
+            AssignmentsQueue assignmentsQueue,
+            boolean isProposedPendingEqualsProposedPlanned,
+            Set<Assignment> partAssignments
+    );
+
+    static class TableGroupUpdateRequestHandler extends GroupUpdateRequestHandler<TablePartitionId> {
+
+        TableGroupUpdateRequestHandler(GroupUpdateRequest request) {
+            super(request);
+        }
+
+        @Override
+        CompletableFuture<Map<TablePartitionId, LocalPartitionStateMessageByNode>> localStatesFuture(
+                DisasterRecoveryManager disasterRecoveryManager,
+                Set<String> zoneNames,
+                Set<Integer> partitionIds,
+                Catalog catalog
+        ) {
+            return disasterRecoveryManager.localPartitionStatesInternal(
+                    zoneNames,
+                    emptySet(),
+                    partitionIds,
+                    catalog,
+                    tableState()
+            );
+        }
+
+        @Override
+        CompletableFuture<Map<Integer, Assignments>> stableAssignments(
+                MetaStorageManager metaStorageManager,
+                int tableId,
+                int[] partitionIds) {
+            return tableStableAssignments(metaStorageManager, tableId, partitionIds);
+        }
+
+        @Override
+        TablePartitionId replicationGroupId(int id, int partitionId) {
+            return new TablePartitionId(id, partitionId);
+        }
+
+        @Override
+        CompletableFuture<Integer> invoke(
```

Comment:
`GroupUpdateRequestHandler.ZoneGroupUpdateRequestHandler#invoke` is almost the same, let's make as less duplicated code as possible. For example,

```
            Iif invokeClosure = prepareMsInvokeClosure(
                    partId,
                    longToBytesKeepingOrder(timestamp.longValue()),
                    assignmentsQueue.toBytes(),
                    // If planned nodes set consists of reset node assignment only then we shouldn't schedule the same planned rebalance.
                    isProposedPendingEqualsProposedPlanned
                            ? null
                            : Assignments.toBytes(partAssignments, assignmentsTimestamp, true)
            );
``` is identical for both cases